vm/interpreter/src/lib.rs

```diff
@@ -1,130 +1,318 @@
-    pub fn new(state: ST, store: &'a DB, epoch: ChainEpoch) -> Self {
+impl<'a, ST: StateTree, DB: BlockStore, SYS: Syscalls> VM<'a, ST, DB, SYS>
+where
+    SYS: Syscalls + Copy,
+{
+    pub fn new(state: ST, store: &'a DB, epoch: ChainEpoch, syscalls: SYS) -> Self {
         VM {
             state,
             store,
             epoch,
+            syscalls,
         }
     }
 
+    /// Returns ChainEpoch
+    fn epoch(&self) -> &ChainEpoch {
+        &self.epoch
+    }
+
     /// Apply all messages from a tipset
     /// Returns the receipts from the transactions.
     pub fn apply_tip_set_messages(
         &mut self,
         _tipset: &Tipset,
         msgs: &TipSetMessages,
-    ) -> Result<Vec<MessageReceipt>, String> {
+    ) -> Result<Vec<ApplyRet>, String> {
         let mut receipts = Vec::new();
 
         for block in &msgs.blocks {
             // TODO: verifiy ordering of message execution
 
             for msg in &block.bls_messages {
-                receipts.push(self.apply_message(msg, &block.miner)?);
+                receipts.push(self.apply_message(msg)?);
             }
 
             for msg in &block.secp_messages {
-                receipts.push(self.apply_message(&msg.message(), &block.miner)?);
+                receipts.push(self.apply_message(msg.message())?);
             }
         }
 
         Ok(receipts)
     }
 
+    fn _apply_implicit_message(&mut self, msg: &UnsignedMessage) -> Result<ApplyRet, String> {
+        let (ret_data, _, act_err) = self.send(msg, 0);
+
+        if let Some(err) = act_err {
+            return Ok(ApplyRet::new(
+                MessageReceipt {
+                    return_data: ret_data,
+                    exit_code: err.exit_code(),
+                    gas_used: BigUint::zero(),
+                },
+                BigUint::zero(),
+                Some(err),
+            ));
+        };
+
+        Ok(ApplyRet::new(
+            MessageReceipt {
+                return_data: ret_data,
+                exit_code: ExitCode::Ok,
+                gas_used: BigUint::zero(),
+            },
+            BigUint::zero(),
+            Some(ActorError::new(ExitCode::Ok, "Ok error".to_owned())),
+        ))
+    }
+
     /// Applies the state transition for a single message
-    /// Returns receipts from the transaction, and the miner penalty token amount.
-    pub fn apply_message(
-        &mut self,
-        msg: &UnsignedMessage,
-        _miner_addr: &Address,
-    ) -> Result<MessageReceipt, String> {
-        check_message(&msg)?;
+    /// Returns ApplyRet structure which contains the message receipt and some meta data.
+    fn apply_message(&mut self, msg: &UnsignedMessage) -> Result<ApplyRet, String> {
+        check_message(msg)?;
 
-        let snapshot = self.state.snapshot()?;
+        let pl = price_list_by_epoch(*self.epoch());
+        let ser_msg = &msg.marshal_cbor().map_err(|e| e.to_string())?;
+        let msg_gas_cost = pl.on_chain_message(ser_msg.len() as i64);
```

Comment:
```suggestion
        let msg_gas_cost = pl.on_chain_message(ser_msg.len()) as u64;
```
Also might make sense to convert to u64 once, instead of doing the conversion throughout this function?